### DISCUSSION TITLE
Custom TURN servers for WebRTC

### DISCUSSION THREAD

**Discussion by AlexBrod-lab** [OvenMediaLabs/OvenMediaEngine#1369] [Help]
Hello,
I want to use OME for WebRTC publish + play (but not only). Typically it's 1 stream and < 3 players, but big amount of such concurrent streams is expected in my case.
But using its own Embedded TURN within OME for video traffic relay will be very costly.
Therefore I'd like to use some other publicly available TURN server(s) with limited, but free bandwidth that I found.

Is it supported at OME that instead of its Embedded TURN, such external TURN server(s) will be provided?
If yes, then how to configure that in both Server.xml and OvenPlayer?

Thanks a lot.

**Comment by basisbit** [comment]
> free TURN servers

Are you sure you aren't referring to STUN servers? I don't know of any `free` TURN servers. Besides that, any such additional server would just cause lots of stuttering and similar playback issues when under load. You'll want the video traffic to directly go from your edge server to the clients.

**Reply by AlexBrod-lab** [reply]
Thanks basisbit,
My idea was actually to offload completely my OME Origin single instance (also no Edge), which is deployed on a tiny, but costly VM on GCP, from the video traffic, which is sent from WebRTC input and consumed using WebRTC OvenPlayer (1-3 viewers per such stream). I still need my own OME for another rare use case with RTMP stream.
So I was thinking how to relay WebRTC streams via another TURN providers, like metered.ca or expressturn.com with their free but generous plans with TURN of 50GB and 1000GB respectively, also their premium plans much cheapest like CPUs/egress traffic on GCP. The STUN signalling is free anyway either via OME or them.
So my question whether it possible to disable embedded TURN on OME and configure another one instead? 
Or better solution might be with more than one TURN servers. If it's possible and supported then how to configure a list of TURN servers on at Server.xml ? And any changes at OvenPlayer in such a case?  ... and can embedded TURN of OME be included as fallback too in such list?

**Reply by basisbit** [reply]
If you want to reduce costs, I'd suggest looking at a virtual server provider like Hetzner or IONOS or Linode.com .
The WebRTC support in OvenMediaEngine only supports direct transport from OME server to the client. OME is not a SFU at all and there is no peer-to-peer traffic between clients, thus your understanding of what the "TURN server" does is not correct in this case.

I suggest either don't use GCP or make use of the Origin->Edge distribution structure which is commonly used for 1-to-N video distribution cases. For a sample configuration, for example look at https://github.com/basisbit/KitchenCDN and also look at https://airensoft.gitbook.io/ovenmediaengine/origin-edge-clustering

**Comment by getroot** [comment]
OME cannot act as a TURN client, but provides TURN server information for players. The role of the built-in TURN server provided by OME is simply to establish an end-to-end TCP connection between the Player and OME.

And even if OME is assumed to be able to perform the role of a TURN client and is assigned a public IP through an external TURN server, that does not mean that OME's traffic can be distributed through the TURN server. Since a channel must be assigned through the TURN server for each Player you want to connect to, the traffic will still be (number of players * bitrates).

**Comment by AlexBrod-lab** [comment]
I got you, thank you guys
